src/CPM/PertMain2.py

```diff
@@ -0,0 +1,405 @@
+        # run from end to start and put all 'ls' 'lf' details in place
+        # calculate slack for all activities (except isolated)
+        # calculate details for isolated activities
+      @ In, None
+      @ Out, None
+    """
+    if self.forwardDict == {}:
+      return
+    self.infoDict[self.startActivity]["ef"] = self.infoDict[self.startActivity]["duration"]
+    self.startToEndScan(self.startActivity)
+    self.infoDict[self.endActivity]["lf"] = self.infoDict[self.endActivity]["ef"]
+    self.infoDict[self.endActivity]["ls"] = self.infoDict[self.endActivity]["lf"] - self.infoDict[self.endActivity]["duration"]
+    self.endToStartScan(self.endActivity)
+    self.calculateSlack()
+    self.generateInfoForIsolated()
+
+  def startToEndScan(self, activity):
+    """
+      Method designed to calculate es and ef of the activities in the schedule
+        # run from start to end and put all 'es' 'ef' details in place
+      @ In, None
+      @ Out, None
+    """
+    for node in self.forwardDict[activity]:
+      if self.infoDict[activity]["ef"] > self.infoDict[node]["es"]:
+        self.infoDict[node]["es"] = self.infoDict[activity]["ef"]
+        self.infoDict[node]["ef"] = self.infoDict[node]["es"] + self.infoDict[node]["duration"]
+      self.startToEndScan(node)
+
+  def endToStartScan(self, activity):
+    """
+      Method designed to calculate ls and lf of the activities in the schedule
+        # run from end to start and put all 'ls' 'lf' details in place
+      @ In, None
+      @ Out, None
+    """
+    for node in self.backwardDict[activity]:
+      if (self.infoDict[node]["lf"] > self.infoDict[activity]["ls"]):
+        self.infoDict[node]["lf"] = self.infoDict[activity]["ls"]
+        self.infoDict[node]["ls"] = (self.infoDict[node]["lf"] - self.infoDict[node]["duration"])
+      self.endToStartScan(node)
+
+  def calculateSlack(self):
+    """
+      Method designed to calculate slack of the activities in the schedule (except isolated)
+      @ In, None
+      @ Out, None
+    """
+    for activity in self.forwardDict:
+      self.infoDict[activity]["slack"] = self.infoDict[activity]["lf"] - self.infoDict[activity]["ef"]
+
+  def generateInfoForIsolated(self):
+    """
+      Method designed to calculate slack for isolated activities
+         # assumption: activity duration shorter than project duration
+      @ In, None
+      @ Out, None
+    """
+    isolated = self.findIsolated()
+    for activity in isolated:
+      self.infoDict[activity]["ef"] = self.infoDict[activity]["es"] + self.infoDict[activity]["duration"]
+      self.infoDict[activity]["lf"] = self.infoDict[self.endActivity]["lf"]
+      self.infoDict[activity]["ls"] = self.infoDict[activity]["lf"] - self.infoDict[activity]["duration"]
+      self.infoDict[activity]["slack"] = self.infoDict[activity]["lf"] - self.infoDict[activity]["ef"]
+
+  # add activity to the pert
+  def addActivity(self, activity, in_connections=[], out_connections=[]):
```

Comment:
camelBack to replace underscore for variables?

Reply:
fixed